add mentor walkthrough steps

ZooeyMiller opened the pull request:
This is being merged to `mentor-walkthrough` not `master` as it only here for this code-along to be consistent in all countries and all future times it is taught. It is for mentors, *not* students os I do not think it should be in master.

This PR adds a series of `step-...` files which contain working examples of how the code should look at all of the different points of the code along. The end of each of these steps is also intended to be a logical point to bring everyone up to the same level in case anyone is lagging behind.

It also adds `node-code-along-walkthrough.md` which is a guide for mentors running the code along that they can study in advanced, and refer to when needed during the code along. 
#7

ZooeyMiller committed:
adds steps and md doc for mentor
this adds the different stages of the node server and a .md file for the mentor to use as a guide for the code-along.
my plan is for this not to be merged into master but left in a branch for the mentors to us for the code-along, but that the repo as it stands is kept clean
#7

ZooeyMiller committed:
adds instructions for step 5
#7

ZooeyMiller committed:
actually adds step-5
didn't save before commiting and pushing. oops.
#7

ZooeyMiller committed:
Merge branch 'master' into updates-for-fac11

ZooeyMiller committed:
deletes .babelrc
I was experimenting with babel in this directory and accidentally pushed up the .babelrc, apologies!
#7

ZooeyMiller committed:
removes babel dependencies
relates to prev commit. apologies.
#7

jsms90 commented:
@ZooeyMiller pulling in some help from @marisid here :+1:

eliasmalik commented on src/handlers/handlers.js:
```diff
@@ -0,0 +1,17 @@
+    path.join(__dirname, '..', '..', 'public', 'index.html'),
+    function(err, file) {
+      if (err) {
+        throw err;
```
Would prefer not to plant the idea of throwing in a handler. I've been commenting during code reviews about this during every cohort. I can see you've used the proper approach in other steps, so perhaps this was an oversight?

ZooeyMiller replied:
Happy to change this

ZooeyMiller replied:
Throws have been changed to the same format as the rest (yes, it was an oversight). 

RE: Solutions branches, if you as maintainers want a different format for solutions I'm happy to acquiesce to that, I was just under the impression that solutions/mentor notes branches were the "done" thing. Hadn't seen foundersandcoders/master-reference#456.

Lock files are part of a node repo. npm explicitly tells you that they are supposed to be committed, my opinion would be that unless we as an organisation explicitly decide they are _not_ to be committed then we should conform to what npm suggest. As for "other stuff" idk what you mean, sorry.

marisid replied:
Inconsistent use of ES5 features (`var`, `function(request, response) {})` which you don't use later - is this on purpose so that you convert to ES6 together in the code-along? I would prefer to stick to ES6 throughout  (which we didn't do for FAC10 as I retrospectively believe we should). The same applies to the router file below.

eliasmalik replied:
Was just including the ref to foundersandcoders/master-reference#456 as reference, not expecting that you'd read it. Anyway, fair to say there's not yet consensus on the branch thing. I'd be happy to merge this into `mentor-walkthrough` for now, I just wanted to raise it with @jsms90 @des-des and @bradreeder.

I agree that lock files are part of a node project. I was asking whether we really needed to make each step-X directory a fully fledged node project; my understanding was that they were just reference examples. If that's true, we'd just need the HTML/CSS/JS. "other stuff" just means the `package.json`, lol.

eliasmalik commented on src/router/router.js:
```diff
@@ -0,0 +1,28 @@
+      ico: 'image/x-icon',
+    };
+    fs.readFile(path.join(__dirname, '..', '..', url), function(err, file) {
+      if (err) throw err;
```
As above, RE: throwing

eliasmalik reviewed (changes requested):
1. As I mentioned foundersandcoders/master-reference#456, I dislike the whole "keep notes/solutions on a different branch" approach. Much prefer having dotfiles or dotfolders for mentor notes. @des-des @bradreeder @jsms90 thoughts?

2. Do we really need the lock-files all the other stuff? Can't we just have the JS/HTML/CSS?

marisid commented on src/server.js:
```diff
@@ -1 +1,24 @@
+
+// to make changes to the page produced by the server we much kill the server and rerun it
+
+// unless you use nodemon
```
Worth highlighting the difference between `start` and `devStart` script tags and the fact that in a development environment you would like to see the changes you make to your app readily, hence `nodemon` is very useful.

ZooeyMiller replied:
Ah, these are actually my original notes from the code-along! :sweat_smile: I should remove them completely, I believe this info is in the guide for mentors .md file

ZooeyMiller committed:
removes throws
as per elias suggestion in https://github.com/foundersandcoders/Node-Intro-Workshop/pull/9#discussion_r127578994

marisid commented on src/server.js:
```diff
@@ -1 +1,24 @@
+
+// you can add shortcuts for your npm modules i nyour package.json
+
+// custom script tags must be run with "npm run" rather than just npm
```
Maybe add here that `install`, `start` and `test` are three common default tags that can be run directly by typing `npm <name-of-script-tag>`

marisid commented on step-1.5/public/index.js:
```diff
@@ -0,0 +1,29 @@
+
+  function httpGetRequest(colour) {
+    var xhr = new XMLHttpRequest();
+    xhr.onreadystatechange = () => {
```
Revert to ES5 `function () {}` instead of ES6 arrow function in front-end JS to avoid issues with browser compatibility. Also, plz make changes to the file in all the steps.

marisid reviewed (changes requested):
Hey @ZooeyMiller great job to add notes - I'm sure this will help you be on track on the day and remind yourself of the next step. Very detailed and descriptive. 

I've now added some suggestions in some parts of the code - plz have a look. 

Also, in general, I feel it's important to keep up with the trends and changes in JS. Great that you've converted the back-end JS in ES6 (with a few exceptions - see my comments) as a lot of great ES6 features are supported in Node. Could I also suggest to introduce `import` and `export default` to the students for importing and exporting modules using ES6 syntax? 

Regarding lock-files, as @eliascodes has mentioned, do we really need them in this workshop?

eliascodes commented:
@marisid ES6modules will require transpilation. Do you think it's worth mentioning it here if they wont use it until (at the very earliest) their week 11 projects (and only if they choose a stack that includes babel)?

ZooeyMiller commented:
@marisid and @eliascodes I really _like_ the idea of using `import` and `export default` and have used it a bit, but I don't feel comfortable with it to the point where I feel I could put it into this workshop and not do it a disservice.

ZooeyMiller committed:
removes files from root src
in line with: https://github.com/foundersandcoders/Node-Intro-Workshop/pull/9#discussion_r127583000

ZooeyMiller committed:
removes front end es6
in line with https://github.com/foundersandcoders/Node-Intro-Workshop/pull/9#discussion_r127582037

ZooeyMiller commented:
All requested changes have been made.
Lock files are still in, but if it's blocking the merge then I can remove them.

marisid commented:
@ZooeyMiller I wouldn't like you to talk about something you don't feel comfortable about. It was just a suggestion, as I think it is good-to-have knowledge when learning Node today. 

@eliascodes even if the students don't use a stack with a transpiler at some point (which I find unlikely in the longer term), I think it would be worth mentioning at the end of the workshop something along the lines: 

*On some occasions you may come across `import`/`export` instead of `require`/`module.exports` in some Node projects. This is basically the equivalent ES6 syntax that can be used in projects where your stack includes a transpiler. A transpiler (such as babel) is a tool to produce equivalent code in another syntax or language.* Which is basically what you've said :smile: 

Again, this is just a suggestion if you feel like the cohort won't be confused and the mentor feels comfortable to deliver.

yvonne-liu reviewed (changes requested):
I would like to fix some markdown issues before this is merged.

yvonne-liu committed:
Update node-code-along-walkthrough.md

yvonne-liu committed:
Update node-code-along-walkthrough.md

yvonne-liu committed:
Update node-code-along-walkthrough.md
